### PR TITLE
Share '$' logic with the normal lexer and the interpolated string scanner.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Continued work to simplify interpolation parsing as precursor to the raw-string-literal interpolations work.

This also moved feature detection of interpolated strings to binding as we were totally inconsistent in parser in terms of how it is done.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs [hunk 1]
```diff
@@ -144,7 +144,7 @@ private ExpressionSyntax ParseInterpolatedStringToken()
             }
 
             Debug.Assert(originalToken.ToFullString() == result.ToFullString()); // yield from text equals yield from node
-            return CheckFeatureAvailability(result, MessageID.IDS_FeatureInterpolatedStrings);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs:147]: this is an example of reverse-smear.  Three syntactic checks becomes one semantic one.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 2]
```diff
@@ -942,6 +935,22 @@ private void ScanSyntaxToken(ref TokenInfo info)
...
+                return true;
+            }
+            else if (TextWindow.PeekChar(1) == '@' && TextWindow.PeekChar(2) == '"')
+            {
+                this.ScanInterpolatedStringLiteral(isVerbatim: true, ref info);
+                return true;
+            }
+
+            return false;
+        }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs:952]: by extracting and sharing the logic here life gets much easier for raw-string literals as there's only one source of truth on determining if we should parse/sub-parse an interpolated string.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @chsienki @jcouv  this is ready for review.  complimentary PR to https://github.com/dotnet/roslyn/pull/57946

**Review by cston — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler for second pair of eyes.  Thanks!

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5)

**Pull request merged**